lms/djangoapps/course_goals/models.py

```diff
@@ -84,3 +93,65 @@ class Meta:
+        '''
+        Update the user activity table with a record for this activity.
+
+        Since we store one activity per date, we don't need to query the database
+        for every activity on a given date.
+        To avoid unnecessary queries, we store a record in a cache once we have an activity for the date,
+        which times out at the end of that date (in the user's timezone).
+
+        The request argument is only used to check if the request is coming from a mobile app.
+        Once the only_if_mobile_app argument is removed the request argument can be removed as well.
+
+        The first return value is the id of the object that was created, or retrieved.
+        A return value of None signifies that there was an issue with the parameters (or the user was masquerading).
+        The second return value indicates whether the object was created or retrieved.
+        '''
+        if not (user and user.id) or not course_key:
+            return None, False
+
+        if only_if_mobile_app and request and not is_request_from_mobile_app(request):
+            return None, False
+
+        if is_masquerading(user, course_key):
+            return None, False
+
+        user_preferences = get_user_preferences(user)
+        timezone = pytz.timezone(user_preferences.get('time_zone', 'UTC'))
+        now = datetime.now(timezone)
+        date = now.date()
+
+        cache_key = 'goals_user_activity_{}_{}_{}'.format(str(user.id), str(course_key), str(date))
+
+        cached_value = TieredCache.get_cached_response(cache_key)
+        if cached_value.is_found:
+            # Temporary debugging log for testing mobile app connection
+            if request:
+                log.info(
+                    'Retrieved cached value with request {} for user and course combination {} {}'.format(
+                        str(request.build_absolute_uri()), str(user.id), str(course_key)
+                    )
+                )
+            return cached_value.value, False
+
+        activity_object, created = cls.objects.get_or_create(user=user, course_key=course_key, date=date)
+
+        # Cache result until the end of the day to avoid unnecessary database requests
+        tomorrow = now + timedelta(days=1)
+        midnight = datetime(year=tomorrow.year, month=tomorrow.month,
+                            day=tomorrow.day, hour=0, minute=0, second=0, tzinfo=timezone)
+        seconds_until_midnight = (midnight - now).seconds
+
+        TieredCache.set_all_tiers(cache_key, activity_object.id, seconds_until_midnight)
+        # Temporary debugging log for testing mobile app connection
+        if request:
+            log.info(
+                'Set cached value with request {} for user and course combination {} {}'.format(
+                    str(request.build_absolute_uri()), str(user.id), str(course_key)
+                )
+            )
+        return (activity_object.id, created)
```

Comment:
Why return the id here? It's never used, and feels like an api oddity - is this call trying to abstract the model or not? Are callers ever going to make further edits to the record row?

This method feels aimed at endpoints that doesn't care about goals or user activity rows - they just want a fire and forget one liner to throw in view code. As a nit only -- this doesn't really matter -- this feels like an `api.py` method more than a model method.

Reply:
hm it feels a bit weird to only return the created value but sure will remove the id

Reply:
hm, it is true that i'm not using the id right now, but if i remove it then there's no way to tell if an object got added or not, which seems awkward?

Reply:
hm, i guess the 200 vs 201 doesn't really matter so i've removed the created argument.
i do like the idea of raising an exception instead of returning None, but i don't want to have to add that in the million files if i'm just going to remove those shortly, but i'm just returning None or the id now and maybe once I remove all the different mobile calls and there will only be 3-4 calls left, I can remove the return value altogether and change it to raise an exception (but i think i can leave that in the meantime so there's at least some way to tell if the method failed)